lib/core/usecase/extend_research_context_usecase.py

```diff
@@ -0,0 +1,149 @@
+    ExtendResearchContextRequest,
+    ExtendResearchContextResponse,
+)
+
+
+class ExtendResearchContextUseCase(ExtendResearchContextInputPort):
+    def execute(
+        self, request: ExtendResearchContextRequest
+    ) -> ExtendResearchContextResponse | ExtendResearchContextError:
+        try:
+            client_repository = self.client_repository
+            research_context_repository = self.research_context_repository
+
+            new_research_context_title = request.new_research_context_title
+            new_research_context_description = request.new_research_context_description
+            client_sub = request.client_sub
+            llm_name = request.llm_name
+            new_source_data_ids_req = request.new_source_data_ids
+            existing_research_context_id = request.existing_research_context_id
+
+            # 1. Get the client, by SUB, to then check if they have access to the source data
+            client_by_sub_dto: GetClientDTO = client_repository.get_client_by_sub(client_sub=client_sub)
+
+            if not client_by_sub_dto.status:
+                return ExtendResearchContextError(
+                    errorCode=client_by_sub_dto.errorCode,
+                    errorMessage=client_by_sub_dto.errorMessage,
+                    errorName=client_by_sub_dto.errorName,
+                    errorType=client_by_sub_dto.errorType,
+                )
+
+            retrieved_client = client_by_sub_dto.data
+
+            if not retrieved_client:
+                return ExtendResearchContextError(
+                    errorCode=404,
+                    errorMessage=f"Client with SUB {client_sub} not found.",
+                    errorName="ClientNotFound",
+                    errorType="ClientNotFound",
+                )
+
+            if not retrieved_client.sub == client_sub:
+                return ExtendResearchContextError(
+                    errorCode=403,
+                    errorMessage=f"Client SUB mismatch! Requested: {client_sub}, Retrieved: {retrieved_client.sub}",
+                    errorName="Client SUB Mismatch",
+                    errorType="ClientSubMismatch",
+                )
+
+            # 2. Check if the client has access to the source data
+            new_client_source_data_list_dto: ListSourceDataDTO = client_repository.list_source_data(
+                client_id=retrieved_client.id
+            )
+
+            if not new_client_source_data_list_dto.status:
+                return ExtendResearchContextError(
+                    errorCode=new_client_source_data_list_dto.errorCode,
+                    errorMessage=new_client_source_data_list_dto.errorMessage,
+                    errorName=new_client_source_data_list_dto.errorName,
+                    errorType=new_client_source_data_list_dto.errorType,
+                )
+
+            new_client_source_data_list = new_client_source_data_list_dto.data
+
+            if not isinstance(new_client_source_data_list, list):
```

Comment:
@maany quick question here: is this check with `isinstance` OK? Don't know if I'm misremembering, but I think you had some opinion against using this kind of check at runtime

Reply:
@ydamit so, the DTO you're getting is a pydantic model (see `ListSourceDataDTO` at `lib/core/dto/client_repository_dto.py`), so pydantic already enforces at runtime that `data` is in fact a list. I.e., we don't need this check with `isinstance`. What you'll need to check instead is that the list is not empty (and thanks to pydantic, if it's not empty, it's guaranteed to be a list of `SourceData` in particular).

Reply:
That's pretty convenient! Do we still need it, then, in the New Research Context feature (which is where I copied it from)? Or is there something about that use case that _does_ require it, that's different here?

Reply:
Oh, my bad, that other one should be a check of a non-empty list too. There's some more cleanup / refactoring to do in the current KP, but we'll tackle that some other day. For now I'll just open a different issue.